packages/utils/src/envelope.ts

```diff
@@ -67,9 +68,24 @@ export function serializeEnvelope(envelope: Envelope, textEncoder?: TextEncoderI
   }
 
   for (const item of items) {
-    const [itemHeaders, payload] = item as typeof items[number];
+    const [itemHeaders, payload] = item;
+
     append(`\n${JSON.stringify(itemHeaders)}\n`);
-    append(typeof payload === 'string' || payload instanceof Uint8Array ? payload : JSON.stringify(payload));
+
+    if (typeof payload === 'string' || payload instanceof Uint8Array) {
+      append(payload);
+    } else {
+      let stringifiedPayload: string;
+      try {
+        stringifiedPayload = JSON.stringify(payload);
+      } catch (e) {
+        // In case, despite all our efforts to keep `payload` circular-dependency-free, `JSON.strinify()` still
+        // fails, we try again after normalizing it again with infinite normalization depth. This of course has a
+        // performance impact but in this case a performance hit is better than throwing.
+        stringifiedPayload = JSON.stringify(normalize(payload));
```

Comment:
Do you think we should throw an error to Sentry as well here? Maybe attach the event_id so folks can report back about why this is happening (like we can inspect the fully normalized payload).

Reply:
It would be nice to log, what parts of the payload contained the circular reference, just so we can close these gaps. We discussed some ideas:
- Capture an additional exception here with the circ-dep path and let users report back to us.
- Set a specific field in `extra` and show a warning on the Sentry product to report back to us.
- Set a specific field in the payload with the circ-dep path and log that field in relay.